File: content/en/04-llms/llms/_solutions/11_parallel/11_parallel.py
# %%
import chatlas

# %%
# Read in the recipes from the text files (this time all of the files)
from pyhere import here

recipe_files = list(here("data/recipes/text").glob("*"))
recipes = [f.read_text() for f in recipe_files]

# %% [markdown]
# We'll use the same Pydantic models we defined in `10_structured-output`.
# Optional: Replace the models in the next cell with your own from that
# exercise.

# %%
from pydantic import BaseModel, Field


class Ingredient(BaseModel):
    name: str = Field(description="Name of the ingredient")
    quantity: str | None = Field(
        description="Quantity as written, including ranges or fractions",
    )
    unit: str | None = Field(
        description="Unit of measure, if applicable",
    )
    notes: str | None = Field(
        description="Additional notes or preparation details",
    )


class Recipe(BaseModel):
    title: str
    description: str
    ingredients: list[Ingredient]
    instructions: list[str] = Field(description="Step-by-step instructions")


# %% [markdown]
# Extract all eight recipes with up to four requests active at once.

# %%
chat = chatlas.ChatPosit(model="claude-haiku-4-5")
results = await chatlas.parallel_chat_structured(
    chat=chat,
    prompts=recipes,
    data_model=Recipe,
    max_active=4,
    on_error="stop",
)
recipes_data = [result.data for result in results]

# %%
[r.title for r in recipes_data]

# %%
# Inspect the extracted recipes in a Polars DataFrame.
import polars as pl

recipes_df = pl.DataFrame([r.model_dump() for r in recipes_data], strict=False)
recipes_df

# %% [markdown]
# Save the recipes, then run `11_recipe-app.R` or `11_recipe-app.py`.

# %%
import json

recipes_structured = [recipe.model_dump() for recipe in recipes_data]

with open(here("data/recipes/recipes.json"), "w") as f:
    json.dump(recipes_structured, f, indent=2)

File: content/en/04-llms/llms/_solutions/11_parallel/11_parallel.R
library(ellmer)

# Read in the recipes from text files (this time all of the recipes)
recipe_files <- fs::dir_ls(here::here("data/recipes/text"))
recipes <- purrr::map(recipe_files, brio::read_file)

# Use the type_recipe we defined in `10_structured-output`.
type_recipe <- type_object(
  title = type_string(),
  description = type_string(),
  ingredients = type_array(
    type_object(
      name = type_string(),
      quantity = type_string(required = FALSE),
      unit = type_string(required = FALSE),
      notes = type_string(required = FALSE)
    )
  ),
  instructions = type_array(type_string())
)

# Extract all eight recipes with up to four requests active at once.
recipes_data <- parallel_chat_structured(
  chat_posit(model = "claude-haiku-4-5"),
  prompts = recipes,
  type = type_recipe,
  max_active = 4,
  on_error = "stop"
)

# Inspect the extracted recipes.
recipes_tbl <- dplyr::as_tibble(recipes_data)
recipes_tbl

# Save the recipes, then run `11_recipe-app.R` or `11_recipe-app.py`.
jsonlite::write_json(
  recipes_data,
  here::here("data/recipes/recipes.json"),
  auto_unbox = TRUE,
  pretty = TRUE
)
